Add Transkribus engine to the OCR tool

kolakachi opened the pull request:
Added Transkribus to the OCR Tool's API
as an option for the engine.

Bug: T326389

kolakachi committed:
Add Transkribus engine to the OCR tool

Added Transkribus to the OCR Tool's API
as an option for the engine.

Bug: T326389

Parthiv-M commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,277 @@
+        if (200 === $statusCode) {
+            $content = json_decode($response->getContent());
+            if (is_null($content)) {
+                $message = 'Transkribus API returned null';
```
I actually changed this error message later on, can we update this here as well? 😅

Parthiv-M commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,277 @@
+        if (200 === $statusCode) {
+            $content = json_decode($response->getContent());
+            if (is_null($content)) {
+                $message = 'Transkribus API returned null';
```
Same thing about the error message here as well

Parthiv-M commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,277 @@
+                $content = json_decode($response->getContent());
+
+                if (is_null($content)) {
+                    $message = 'Transkribus API returned null';
```
- Need to modify the error message again
- I think this is missing a call to `setErrorMessage()`

Parthiv-M reviewed (dismissed):
Need to change the auth flow once the CLI command is complete

samwilson commented on config/services.yaml:
```diff
@@ -41,6 +41,12 @@ services:
+    App\Engine\TranskribusEngine:
+        parent: App\Engine\EngineBase
+        arguments:
+            $username: '%env(USERNAME)%'
```
This should only be reading in the `APP_TRANSKRIBUS_ACCESS_TOKEN` from `.env` and not the username and password. PR #63 needs to be merged first, of course.

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -0,0 +1,86 @@
+            'username' => $username,
+            'password' => $password
+        ];
+        $this->transkribusClient = new TranskribusClient($httpClient, $credentials);
```
The TranskribusClient object shouldn't be created here, it should be injected as a constructor parameter. That way, the service container handles instantiation.

samwilson commented on templates/output.html.twig:
```diff
@@ -23,6 +23,12 @@
+                <div class="radio">
+                    <label>
+                        <input type="radio" name="engine" value="transkribus" {% if engine == 'transkribus' %}checked{% endif %} />
+                       Transkribus OCR
```
Needs internationalization (i.e. using `msg()` and adding to `en.json` and `qqq.json`). Copy how the others are handled.

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -0,0 +1,86 @@
+        $response = $this->transkribusClient->initProcess($imageUrl, []);
+
+        if ($response->getHasError()) {
+            file_put_contents('initp.json', "init process failed");
```
What is this file for? It doesn't seem to be used anywhere.

kolakachi replied:
Will remove that, was trying to debug

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,277 @@
+            $data = file_get_contents('transkribus_credentials.json');
+            $content = json_decode($data);
+            $refreshToken = $content->{'refresh_token'};
+            $response = $this->httpClient
```
The pattern of making a request, looking at its response code and content, and returning errors, is repeated a few times. It's probably worth abstracting into a separate private function e.g. `TranskribusClient::request()`. That way, the TranskribusAuthCommand could also have TranskribusClient injected and use it to make the requests there.

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,277 @@
+        }        
+    }
+
+    private function retrieveAccessToken(){
```
The authentication process will be done by #63 so doesn't need to be repeated here.

samwilson reviewed (commented):
This is looking good! I've not gone through it deeply, but just made few comments about things that I noticed. The main thing now I think is to get #63 merged, and then rebase this on main so it doesn't have to do any of the OAuth stuff. Then we can look into the new engine workflow properly.

samwilson commented:
Oh, and the other thing: make sure you run `composer test` and `npm run test` before pushing. That'll catch some of the simpler things.

kolakachi committed:
Merge branch 'main' into add-transkribus-engine

kolakachi committed:
Inject transkribus as a constructor parameter

kolakachi committed:
fix formatting error

kolakachi committed:
fix line spacing error

kolakachi committed:
Fix formatting error

Refactored TranskribusClient to recieve only image url

kolakachi committed:
Update Transkribus engine internationalization

using msg() in output.html.twig, created internationalization by
adding Transkribus to en.json and qqq.json

kolakachi committed:
Create a method to serve all Transkribus request

Refactored TranskribusClient to handle all request through
a method

kolakachi committed:
Change Transkribus request method to public

kolakachi committed:
Fix type hinting errors

kolakachi committed:
specify type hint for items of parameter jsonBody

kolakachi committed:
specify array<string,mixed> parameter jsonBody

kolakachi committed:
specify mixed[] parameter jsonBody

samwilson commented on config/services.yaml:
```diff
@@ -41,6 +41,13 @@ services:
+    
+    App\Engine\TranskribusEngine:
+        parent: App\Engine\EngineBase
+            
```
Trailing spaces here and above on line 44.

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,179 @@
+    private $config = [];
+
+    /** @var $langs valid languages. */
+    private $langs = '';
```
Not used.

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -0,0 +1,85 @@
+class TranskribusEngine extends EngineBase
+{
+    /** @var string The API key. */
+    protected $key;
```
Not used.

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -0,0 +1,85 @@
+        HttpClientInterface $httpClient
+    ) {
+        parent::__construct($intuition, $projectDir, $httpClient);
+        
```
Trailing spaces.

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,179 @@
+    /**
+     * @param string $imageURL
+     */
+    public function initProcess(string $imageURL): self
```
Why is this a fluent method? (i.e. returns itself). I would suggest that this should either throw an exception in case of an error, or return the process ID (or whatever else is required).

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,179 @@
+        return $this;
+    }
+
+    public function retrieveProcessStatus(): self
```
This should take a process ID as a parameter, and return the status of that process.

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,179 @@
+     * @param mixed[] $jsonBody
+     * @return object
+     */
+    public function request(
```
Doesn't need to be public. Users of this class shouldn't be able to make requests directly.

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,179 @@
+
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+class TranskribusClient
```
This class shouldn't be mutable like this. At the moment, it holds response info, error messages etc. that relate to the most recently called process. This is fragile, and makes the class harder to understand from the outside. This class is responsible for being the interface to the remote API, and should make it unnecessary for people using this class to know anything about that API. It looks like it needs something like the following:

1. `initProcess(string $imageUrl): string` to get a process ID;
2. `retrieveProcess(string $processId): array` to get the process result or current status;

Could possibly also have the authentication-related stuff, but that could be a subsequent improvement:

3. `authInit($username, $password)`
4. `authRefresh($token)`

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -0,0 +1,85 @@
+
+        $image = $this->getImage($imageUrl, $crop);
+        $imageUrl = $image->getUrl();
+        $response = $this->transkribusClient->initProcess($imageUrl);
```
This is not a response object here, it's still a TranskribusClient, so I'd suggest making the name clearer. However, I think the fluent style is not applicable here and perhaps shouldn't be used.

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -0,0 +1,85 @@
+        $response = $this->transkribusClient->initProcess($imageUrl);
+
+        if ($validLangs) {
+            $response->setLanguages($this->getLangCodes($validLangs));
```
The languages are being set here on the client object *after* the request has been sent.

The whole topic of languages/models I think could probably wait for a separate patch anyway, because it does complicate things, and you've not implemented it in this patch yet anyway.

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -0,0 +1,85 @@
+        while (!$response->hasError() && '' === $response->getTextResult() && 11 >$counter) {
+            $response->retrieveProcessStatus();
+            $counter++;
+            sleep(5);
```
I think a wait of up to 55 seconds is probably too much. But we can revisit this later I'm sure.

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        $counter = 0;
+        while (!$response->hasError() && '' === $response->getTextResult() && 11 >$counter) {
```
A result of empty string is valid and shouldn't be a loop condition. Is it not enough to look for error messages or response status?

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,179 @@
+        $this->langs = $langs;
+    }
+
+     /**
```
There's an extraneous space here and on line 134.

samwilson reviewed (commented):
A few comments on various things. Mainly I think the immutability of the client class is the most important.

kolakachi committed:
Refactor Transkribus upload process

Refactored class TranskribusClient and class TranskribusEngine

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,150 @@
+     * @param mixed[] $config
+     * @return int
+     */
+    public function initProcess(string $imageURL, array $config = []): int
```
Is `$config` ever used? It looks like we might want to pass an `$htrId` here instead, and use that in the config (which wouldn't need to be conditionally constructed).

kolakachi replied:
Noted, I think we need a parameter for content. This holds image co-ordinates.

samwilson replied:
That's true, but it's not being used yet so there's no need to add it yet. It can be refactored as required.

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,150 @@
+    /** @var $config. */
+    private $config = [];
+
+    /** @var string Transcribed result. */
```
Wrong comment (this is the API URL, not a result).

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,150 @@
+    private $httpClient;
+
+    /** @var string transkribus access token. */
+    private $accessToken = '';
```
No need for a default value here, as this will always be set in the constructor so will never actually have this value.

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,150 @@
+
+            if (is_null($content)) {
+                $message = $this->getErrorMessage(0);
+                throw new OcrException('transkribus-error', [$message]);
```
The `transkribus-error` message needs to be defined in `en.json` and `qqq.json`.

kolakachi replied:
I noticed The API has different error messages and status codes, is it advisable to let Tranksribus define the error messages? or do we go along with defining the messages ourselves?

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,150 @@
+        $content = $this->request('POST', self::PROCESSES_URL, $jsonBody);
+        $parsedContent = (array) $content;
+        if (!empty($parsedContent)) {
+            if ($content->{'status'} === 'CREATED') {
```
There are five places where you're referring to member variables with braces where these don't contain illegal characters. These can all be simplified to e.g. `$content->status`.

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,150 @@
+        $url = self::PROCESSES_URL.'/'.$processId;
+
+        $content = $this->request('GET', $url, []);
+        $parsedContent = (array) $content;
```
Why are you casting to an array?

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,150 @@
+        if (!empty($parsedContent)) {
+            if ($content->{'status'} === 'FINISHED') {
+                $textContent = $content->{'content'};
+                $textResult = $textContent->{'text'};
```
This could be simplified to `$textResult->content->text ?? null;`

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,150 @@
+        $textResult = '';
+
+        if (!empty($parsedContent)) {
+            if ($content->{'status'} === 'FINISHED') {
```
How are you distinguishing between a finished process and one that returns an empty string? If we call `retrieveProcessResult()` and get an empty string, we don't know whether to call it again or not.

kolakachi replied:
Currently checking for FINISHED and FAILED. The code currently covers CREATED, WAITING and RUNNING.

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -0,0 +1,69 @@
+        $imageUrl = $image->getUrl();
+        $processId = $this->transkribusClient->initProcess($imageUrl);
+
+        if (0 === $processId) {
```
The process ID of `0` is hard coded in the client. I suggest that initProcess instead should throw an exception if no process ID was returned.

samwilson commented on templates/output.html.twig:
```diff
@@ -23,6 +23,12 @@
+                        <input type="radio" name="engine" value="transkribus" {% if engine == 'transkribus' %}checked{% endif %} />
+                        {{ msg('engine-name-transkribus') }}
+                    </label>
+                </div>
```
This UI is pretty simple, so it's good to have it here. But note that this task is about adding to the API so should also have additional edits to the OpenApi docs for the `apiAction()` method in the controller.

Parthiv-M commented on src/Engine/TranskribusClient.php:
```diff
@@ -0,0 +1,150 @@
+        $errorMessage = '';
+        switch ($statusCode) {
+            case 0:
+                $errorMessage = 'Transkribus API returned null';
```
Should we be changing this message to something like `Could not parse result from Transkribus API` instead?

kolakachi committed:
Refactor Transkribus Enigine and Client

Defined transkribus error messages in en.json and qqq.json
updated initProcess and retrieveProcessResult in TranskribusClient

kolakachi committed:
Update OpenApi docs for the apiAction method

kolakachi committed:
Fix sniff violations

Parthiv-M commented on src/Controller/OcrController.php:
```diff
@@ -184,7 +185,7 @@ public function homeAction(): Response
      *     name="engine",
      *     in="query",
-     *     description="The engine to use, either `tesseract` or `google`.",
+     *     description="The engine to use, either `tesseract` or `google` or transkribus.",
```
I think it should be changed to \`transkribus`, so that it conforms to the rest of the engine name styles

Parthiv-M commented:
Also, the `description` field [here](https://github.com/wikimedia/wikimedia-ocr/blob/2b7ec341a505f23c04e66650053a46cf614a4d1d/config/packages/nelmio_api_doc.yaml#L5) also needs to be changed to include Transkribus

kolakachi committed:
update api doc description

samwilson commented on i18n/qqq.json:
```diff
@@ -56,5 +57,10 @@
 	"tesseract-param-error": "Error message displayed when invalid values for Tesseract options are submitted.\n\nParameters:\n* $1 – the form label for the option. Currently, the only possible value here is {{msg-wm|wikimedia-ocr-tesseract-psm-label}}.\n* $2 – The value that was given.\n* $3 – the maximum value for the option (this will be an integer).",
-	"tesseract-internal-error": "Generic error message displayed when the tesseract command fails."
+	"tesseract-internal-error": "Generic error message displayed when the tesseract command fails.",
+	"transkribus-error-401": "Error Code '$1' :: The request is not authorized",
```
The sentences in qqq.json are the help text to go along with the actual messages in en.json. They're shown to translators on TranslateWiki.net, who will use them to make decisions about how to translate things. So the need to explain what the message is for. See https://www.mediawiki.org/wiki/Help:System_message for some more info (this tool uses the pretty much the same system as MediaWiki).

samwilson reviewed (commented):
Sorry, I was just about to merge this but I noticed that the message documentation isn't correct. (As soon as patches are merged, their i18n data is updated on translatewiki.net, so we can't do this in a follow-up commit.)

kolakachi committed:
update transkribus error translations
Changed sentences in qqq.json to describe error texts in en.json
Updated OcrException i18nkeys in TranskribusClient
